lib/ssl_compat/ssl_compat.c

```diff
@@ -0,0 +1,121 @@
+#include <assert.h>
+#include <stddef.h>
+
+#include <openssl/ssl.h>
+
+#include "ssl_compat.h"
+
+#ifndef OPENSSL_VERSION_NUMBER
+#error "OPENSSL_VERSION_NUMBER must be defined"
+#endif
+
+/*
+ * LibreSSL claims to be OpenSSL 2.0, but (currently) has APIs compatible with
+ * OpenSSL 1.0.1g.
+ */
+#ifdef LIBRESSL_VERSION_NUMBER
+#undef OPENSSL_VERSION_NUMBER
+#define OPENSSL_VERSION_NUMBER 0x1000107fL
+#endif
+
+/**
+ * ssl_compat_write_ex(ssl, buf, num, written):
+ * Write ${num} bytes from ${buf} to the ${ssl} connection.  Store the number
+ * of bytes written in ${written}.
+ *
+ * On success, return 1.  On failure, return 0 or a negative value (for some
+ * versions of OpenSSL, the exact value must be passed to SSL_get_err() in
+ * order to receive a meaningful error code).
+ *
+ * COMPATIBILITY: if you are using OpenSSL < 1.1.1 or libressl, and give a
+ * value of ${num} > INT_MAX, this function will silently use a value of
+ * INT_MAX instead.
+ */
+int
+ssl_compat_write_ex(SSL * ssl, const void * buf, size_t num, size_t * written)
+{
+
+	/* Sanity check. */
+	assert(num > 0);
+
+#if OPENSSL_VERSION_NUMBER < 0x10101000L
+	int ret;
+
+	/* Reduce the number of bytes to write (if necessary). */
+	if (num > INT_MAX)
+		num = INT_MAX;
+
+	/* Attempt to send data. */
+	ret = SSL_write(ssl, buf, (int)num);
+
+	if (ret > 0) {
+		/* Sanity check. */
+		assert(ret <= (int)num);
+
+		/* Record the number of bytes written, and overall success. */
+		*written = (size_t)ret;
+		ret = 1;
+	} else {
+		/*
+		 * Do nothing here, because ret is a meaningful value for
+		 * determining the error.
+		 */
+	}
+
+	return (ret);
+#else
```

Comment:
My code had the `#else` paths for "if we don't need to use this compat code".  Now that we're replacing the original functions with `#define SSL_read_ex ssl_compat_read_ex`, I'm not at all certain that we want to keep them.

I was thinking that it might be useful to compare / verify the "compat" portion by seeing the new function(s).  In this case, it's trivial, but for openssl 1.0.2 it gets a lot messier.  Maybe we should just keep this question in the backs of our minds, and re-evaluate once I upload the 1.0.2 version.